packages/context/src/context.ts

```diff
@@ -797,27 +798,81 @@ export class Context extends EventEmitter {
    */
   // TODO(rfeng): Evaluate https://nodejs.org/api/util.html#util_custom_inspection_functions_on_objects
   inspect(options: ContextInspectOptions = {}): JSONObject {
+    return this._inspect(options, new ClassNameMap());
+  }
+
+  /**
+   * Inspect the context hierarchy
+   * @param options - Options for inspect
+   * @param visitedClasses - A map to keep class to name so that we can have
+   * different names for classes with colliding names. The situation can happen
+   * when two classes with the same name are bound in different modules.
+   */
+  private _inspect(
+    options: ContextInspectOptions,
+    visitedClasses: ClassNameMap,
+  ): JSONObject {
     options = {
       includeParent: true,
       includeInjections: false,
       ...options,
     };
     const bindings: JSONObject = {};
     for (const [k, v] of this.registry) {
+      const ctor = v.valueConstructor ?? v.providerConstructor;
+      let name: string | undefined = undefined;
+      if (ctor != null) {
+        name = visitedClasses.visit(ctor);
+      }
       bindings[k] = v.inspect(options);
+      if (name != null) {
+        const binding = bindings[k] as JSONObject;
+        if (v.valueConstructor) {
+          binding.valueConstructor = name;
+        } else if (v.providerConstructor) {
+          binding.providerConstructor = name;
+        }
+      }
     }
     const json: JSONObject = {
       name: this.name,
       bindings,
     };
     if (!options.includeParent) return json;
     if (this._parent) {
-      json.parent = this._parent.inspect(options);
+      json.parent = this._parent._inspect(options, visitedClasses);
     }
     return json;
   }
 }
 
+/**
+ * An internal utility class to handle class name conflicts
+ */
+class ClassNameMap {
+  private readonly classes = new Map<Constructor<unknown>, string>();
+  private readonly nameIndex = new Map<string, number>();
+
+  visit(ctor: Constructor<unknown>) {
+    let name = this.classes.get(ctor);
+    if (name == null) {
+      name = ctor.name;
+      // Now check if the name collides with another class
+      let index = this.nameIndex.get(name);
+      if (typeof index === 'number') {
+        // A conflict is found, mangle the name
+        this.nameIndex.set(name, ++index);
```

Comment:
The resulting names look a lot like the names of nested java classes.  `Myprovider$1` .

For bindings, you gave them an unique id: `Binding_2_3`  for example.
Can we do the same?

Reply:
Binding ids are not generated by `inspect()`.

I was trying to use `_1` but there is a possibility that a class named `MyProvider_1` may exist. I don't want to make the name mangling too complex.

Reply:
I see ok.